### PR TITLE
Connect4 trial dontjudgeme

### PR DESCRIPTION
**Pull request by Razvi28**
Pls tell me it works:))

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,3 +1,10 @@
+extern crate tokio;
```
Comment (konsolas) [inline src/main.rs:1]: Don't think these lines need to be added, please remove them.

File: src/main.rs [hunk 2]
```diff
@@ -27,6 +35,7 @@ async fn main() -> std::io::Result<()> {
             .service(api::get_state)
             .service(api::submit_move)
             .service(api::wait_for_update)
+
```
Comment (konsolas) [inline src/main.rs:38]: Please revert the unnecessary changes in main.rs

File: src/game/connect4.rs [hunk 3]
```diff
@@ -2,27 +2,46 @@ use crate::game::adapter::{
     GameAdapter, GameAdapterError, GenericGameMove, GenericGameState, State,
 };
 use crate::game::{GameId, SessionId};
+
+use actix_web::client::SendRequestError;
```
Comment (konsolas) [inline src/game/connect4.rs:6]: Unused import

File: src/game/connect4.rs [hunk 4]
```diff
@@ -2,27 +2,46 @@ use crate::game::adapter::{
...
+use actix_web::client::WsClientError::SendRequest;
```
Comment (konsolas) [inline src/game/connect4.rs:7]: Unused import

File: src/game/connect4.rs [hunk 5]
```diff
@@ -65,20 +89,281 @@ impl GameAdapter for Connect4Adapter<'_> {
...
 
-impl Connect4<'_> {}
+impl Connect4 {
+    fn get_cell_at(&self, row: usize, col: usize) -> Option<Token> {
+        Some(*self.board.get(col)?.get(row)?)
+    }
+
+    fn insert_move_if_legal(&mut self, column: usize) -> actix_web::Result<()> {
+        if column >= COL_SIZE {
+            return Err(actix_web::Error::from(GameAdapterError::InvalidGameState(
```
Comment (konsolas) [inline src/game/connect4.rs:156]: Is this really an "invalid game state" error? The game state "in progress" is valid - it's the *move* that is invalid.

File: src/game/connect4.rs [hunk 6]
```diff
@@ -65,20 +89,281 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+                lenl += 1;
+                col_aux += 1;
+                row_aux += 1;
+            }
+            lenl + lenr >= CONNECT_FOUR
+        } else {
+            false
+        }
+    }
+    fn draw_move(&self) -> bool {
```
Comment (konsolas) [inline src/game/connect4.rs:302]: Would prefer a name like `is_game_drawn` - don't think draw_move really captures what this method is doing.

File: src/game/connect4.rs [hunk 7]
```diff
@@ -65,20 +89,281 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+    fn moves(&mut self, column: usize) -> actix_web::Result<()> {
+        if self.completed {
+            return Err(actix_web::Error::from(GameAdapterError::InvalidGameState(
+                State::Ended,
+            )));
+        }
+        self.insert_move_if_legal(column)?;
+        Ok(())
+    }
+    fn visualize(&mut self, column: usize) {
```
Comment (konsolas) [inline src/game/connect4.rs:315]: This is not used outside of testing so it should have the `#[cfg(test)]` attribute

File: src/game/connect4.rs [hunk 8]
```diff
@@ -65,20 +89,281 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+        };
+        board1.visualize(8);
+        board1.visualize(2);
+        board1.visualize(2);
+        board1.visualize(2);
+        board1.visualize(2);
+        board1.visualize(2);
+        board1.visualize(2);
+
+        assert_eq!(2 + 2, 4);
```
Comment (konsolas) [inline src/game/connect4.rs:367]: Should probably use assertions to check that things worked correctly rather than relying on a human to read the output manually

**Review by konsolas — CHANGES_REQUESTED** [review]
Haven't looked at all of the code in detail yet but here are a few things. Will wait for @jonathanjameswatson and @MilkFansHelloWorld to review as well.

File: src/game/mod.rs [hunk 9]
```diff
@@ -176,8 +178,10 @@ impl GameManager {
             sessions: DashMap::new(),
         }
     }
-
-    pub fn create_game(&self, game: impl FnOnce(GameId) -> Box<dyn GameAdapter>) -> Result<GameId> {
+    pub fn create_game(
+        &mut self,
+        game: impl FnOnce(GameId) -> Box<dyn GameAdapter>,
+    ) -> Result<GameId> {
```
Comment (jonathanjameswatson) [inline src/game/mod.rs:184]: Please do not mutably borrow self here — it is not necessary and it stops the project building

File: src/game/connect4.rs [hunk 10]
```diff
@@ -65,20 +89,281 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+                self.game.completed = true;
+            } else {
+                self.game.switch_token();
+            }
+            Ok(())
+        } else {
+            Err(actix_web::Error::from(
+                GameAdapterError::WrongPlayerRequest(user), // return the one who made the request
+            ))
+        }
```
Comment (jonathanjameswatson) [inline src/game/connect4.rs:121]: Invert this if statement to reduce nesting

**Review by jonathanjameswatson — CHANGES_REQUESTED** [review]
Looks very good apart from some style inconsistencies and some unnecessary changes

File: src/game/connect4.rs [hunk 11]
```diff
@@ -42,18 +62,17 @@ impl GameAdapter for Connect4Adapter<'_> {
...
-            Err(actix_web::Error::from(
+        if self.players.len() >= NUM_PLAYERS {
+            return Err(actix_web::Error::from(
                 GameAdapterError::PlayerLimitExceeded(NUM_PLAYERS),
-            ))
+            ));
+        }
+        self.players.push(username);
+        if self.players.len() == NUM_PLAYERS {
+            self.state = State::InProgress
```
Comment (konsolas) [inline src/game/connect4.rs:72]: Need to set `self.next_move`

File: src/game/connect4.rs [hunk 12]
```diff
@@ -65,20 +84,224 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+                Token::Blue => self.players.get(1).unwrap().clone(),
+            };
+            self.winner.push(winner);
+        }
+        if win || draw {
+            self.game.completed = true;
+        } else {
+            self.game.switch_token();
+        }
+        Ok(())
```
Comment (konsolas) [inline src/game/connect4.rs:116]: Need to send an update notification with `self.notifier.send(());`

File: src/game/connect4.rs [hunk 13]
```diff
@@ -65,20 +84,224 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+            let winner = match self.game.turn {
+                Token::Red => self.players.get(0).unwrap().clone(),
+                Token::Blue => self.players.get(1).unwrap().clone(),
+            };
+            self.winner.push(winner);
+        }
+        if win || draw {
+            self.game.completed = true;
+        } else {
+            self.game.switch_token();
```
Comment (konsolas) [inline src/game/connect4.rs:114]: And update `self.next_move`

**Review by konsolas — CHANGES_REQUESTED** [review]
This is all that stands out - lmk if any questions

File: src/game/connect4.rs [hunk 14]
```diff
@@ -65,20 +84,224 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+                        Token::Blue => self.players.get(1),
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
         Ok(GenericGameState {
             game: "connect_4".to_string(),
             players: self.players.clone(),
             state: self.state,
             can_move: vec![self.next_move.clone()],
```
Comment (jonathanjameswatson) [inline src/game/connect4.rs:137]: Instead of having `self.next_move` in your code, just convert `turn` into the first or second player depending on if it is red or blue (unless the game hasn't started yet or it has ended, in which case just have `can_move` be an empty vector)

**Review by jonathanjameswatson — COMMENTED** [review]
Almost there now

File: src/game/connect4.rs [hunk 15]
```diff
@@ -42,43 +60,249 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+            })
+            .collect::<Vec<_>>();
+
         Ok(GenericGameState {
             game: "connect_4".to_string(),
             players: self.players.clone(),
             state: self.state,
-            can_move: vec![self.next_move.clone()],
+            can_move: vec![self.get_user_from_token()],
             winners: vec![],
```
Comment (jonathanjameswatson) [inline src/game/connect4.rs:139]: Make sure to put `self.winner` in here

File: src/game/connect4.rs [hunk 16]
```diff
@@ -42,43 +60,249 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+
         Ok(GenericGameState {
             game: "connect_4".to_string(),
             players: self.players.clone(),
             state: self.state,
-            can_move: vec![self.next_move.clone()],
+            can_move: vec![self.get_user_from_token()],
```
Comment (jonathanjameswatson) [inline src/game/connect4.rs:138]: Do not call this if the game state isn't `InProgress`, just return an empty vector instead

File: src/game/connect4.rs [hunk 17]
```diff
@@ -42,43 +60,249 @@ impl GameAdapter for Connect4Adapter<'_> {
...
             winners: vec![],
-            payload: serde_json::to_value(&self.game.board)?,
+            payload: serde_json::to_value(&encoded_board)?,
```
Comment (jonathanjameswatson) [inline src/game/connect4.rs:140]: Payload should be a JSON object containing a list rather than a list

File: src/game/connect4.rs [hunk 18]
```diff
@@ -42,43 +60,249 @@ impl GameAdapter for Connect4Adapter<'_> {
...
+        self.game.moves(column)?;
+        let win = self.game.winning_move(column);
+        let draw = self.game.is_game_drawn();
+        if win {
+            self.winner.push(self.get_user_from_token());
+        }
+        if win || draw {
+            self.game.completed = true;
+        } else {
+            self.game.switch_token();
```
Comment (jonathanjameswatson) [inline src/game/connect4.rs:113]: Make sure to set the game state to ended as well

**Review by jonathanjameswatson — CHANGES_REQUESTED** [review]
Here are some fixed that are required after debugging

File: src/game/connect4.rs [hunk 19]
```diff
@@ -42,43 +65,260 @@ impl GameAdapter for Connect4Adapter<'_> {
     }
 
     fn add_player(&mut self, username: String) -> actix_web::Result<()> {
-        if self.players.len() < NUM_PLAYERS {
-            self.players.push(username);
-            if self.players.len() == NUM_PLAYERS {
-                self.state = State::InProgress
-            }
```
Comment (jonathanjameswatson) [inline src/game/connect4.rs:49]: We need these lines to change the state when the game starts

**Review by jonathanjameswatson — CHANGES_REQUESTED** [review]
Almost works now

**Review by jonathanjameswatson — APPROVED** [review]
I am happy with this

File: src/game/adapter.rs [hunk 20]
```diff
@@ -12,6 +12,9 @@ use tokio::sync::broadcast;
 pub enum GameAdapterError {
     PlayerLimitExceeded(usize),
     InvalidGameState(State),
+    WrongPlayerRequest(String),
+    WrongMoveRequest(usize),
+    GameEnded(),
```
Comment (konsolas) [inline src/game/adapter.rs:17]: isn't this a special case of `InvalidGameState(Ended)`?

**Review by konsolas — APPROVED** [review]
Happy to merge as is, but lmk what you think about this.

File: src/game/adapter.rs (hunk 20 above)
Comment (Razvi28) [reply src/game/adapter.rs:17]: Yeah it is, I was talking to Johnathan and he told me we should print a different message when sb tries to make a move after the game has ended, I changed it accordingly.

If u want i can revert it back the way it was, without that special case
